src/security/context.cpp

```diff
@@ -0,0 +1,672 @@
+#include "context.h"
+
+#include <ngx_core.h>
+#include <ngx_http_core_module.h>
+#include <ngx_log.h>
+
+#include <atomic>
+#include <charconv>
+#include <optional>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <type_traits>
+#include <unordered_set>
+#include <utility>
+#include <variant>
+
+#include "../datadog_conf.h"
+#include "../datadog_context.h"
+#include "../datadog_handler.h"
+#include "../ngx_http_datadog_module.h"
+#include "../tracing_library.h"
+#include "blocking.h"
+#include "collection.h"
+#include "ddwaf_obj.h"
+#include "header_tags.h"
+#include "library.h"
+#include "util.h"
+
+extern "C" {
+#include <ngx_hash.h>
+#include <ngx_http.h>
+#include <ngx_regex.h>
+#include <ngx_string.h>
+#include <ngx_thread_pool.h>
+}
+
+#include <datadog/span.h>
+#include <datadog/span_data.h>
+#include <datadog/trace_segment.h>
+#include <ddwaf.h>
+#include <rapidjson/document.h>
+#include <rapidjson/encodings.h>
+#include <rapidjson/prettywriter.h>
+
+using namespace std::literals;
+
+namespace {
+
+namespace dnsec = datadog::nginx::security;
+
+class JsonWriter : public rapidjson::Writer<rapidjson::StringBuffer> {
+  using rapidjson::Writer<rapidjson::StringBuffer>::Writer;
+
+ public:
+  // NOLINTNEXTLINE(readability-identifier-naming)
+  bool ConstLiteralKey(std::string_view sv) {
+    return String(sv.data(), sv.length(), false);
+  }
+};
+
+void ddwaf_object_to_json(JsonWriter &w, const ddwaf_object &dobj);
+
+void report_match(const ngx_http_request_t &req, dd::TraceSegment &seg,
+                  dd::Span &span,
+                  std::vector<dnsec::OwnedDdwafResult> &results) {
+  if (results.empty()) {
+    return;
+  }
+
+  seg.override_sampling_priority(2);  // USER-KEEP
+  span.set_tag("appsec.event"sv, "true");
+
+  rapidjson::StringBuffer buffer;
+  JsonWriter w(buffer);
+  w.StartObject();
+  w.ConstLiteralKey("triggers"sv);
+
+  w.StartArray();
+  for (auto &&result : results) {
+    auto events = dnsec::ddwaf_arr_obj{(*result).events};
+    for (auto &&evt : events) {
+      ddwaf_object_to_json(w, evt);
+    }
+  }
+  w.EndArray(results.size());
+
+  w.EndObject(1);
+  w.Flush();
+
+  std::string_view const json{buffer.GetString(), buffer.GetLength()};
+
+  ngx_str_t json_ns{dnsec::ngx_stringv(json)};
+  ngx_log_error(NGX_LOG_NOTICE, req.connection->log, 0, "appsec event: %V",
```

Comment:
Any particular reason to choose notice instead of debug? I'm not sure what the typical setup will be but this is a pretty verbose log statement.

Reply:
Notice is between info and warn. So maybe info is better instead? Debug is not a great option, because it's compiled to nothing by default. You need to configure nginx with --debug and then build the extension against that, which we don't do (and I doubt the result would be binary compatible, though I didn't check).

Reply:
changed to info